File: Source/AlphaTab/Rendering/Glyphs/GroupedEffectGlyph.cs
using AlphaTab.Model;
using AlphaTab.Platform;

namespace AlphaTab.Rendering.Glyphs
{
    abstract class GroupedEffectGlyph : EffectGlyph
    {
        protected BeatXPosition EndPosition { get; set; }
        protected bool ForceGroupedRendering { get; set; }
        protected bool EndOnBarLine { get; set; }

        protected GroupedEffectGlyph(BeatXPosition endPosition) : base(0, 0)
        {
            EndPosition = endPosition;
        }

        /// <summary>
        /// Gets a value whether this glyph is linked with a previous glyph for rendering. 
        /// This means this glyph will not be rendered itself, but rendered as part of the very first glyph of this link-group.
        /// </summary>
        public bool IsLinkedWithPrevious
        {
            get
            {
                return PreviousGlyph != null && PreviousGlyph.Renderer.Staff.StaveGroup == Renderer.Staff.StaveGroup;
            }
        }

        /// <summary>
        /// Gets a value whether this glyph is linked with the next glyph for rendering. 
        /// </summary>
        public bool IsLinkedWithNext
        {
            get
            {
                // we additionally check IsFinalized since the next renderer might not be part of the current partial
                // and therefore not finalized yet. 
                return NextGlyph != null && NextGlyph.Renderer.IsFinalized && NextGlyph.Renderer.Staff.StaveGroup == Renderer.Staff.StaveGroup;
            }
        }

        public override void Paint(float cx, float cy, ICanvas canvas)
        {
            // if we are linked with the previous, the first glyph of the group will also render this one.
            if (IsLinkedWithPrevious)
            {
                return;
            }

            // we are not linked with any glyph therefore no expansion is required, we render a simple glyph. 
            if (!IsLinkedWithNext && !ForceGroupedRendering)
            {
                PaintNonGrouped(cx, cy, canvas);
                return;
            }

            // find last linked glyph that can be  
            GroupedEffectGlyph lastLinkedGlyph;
            if (!IsLinkedWithNext && ForceGroupedRendering)
            {
                lastLinkedGlyph = this;
            }
            else
            {
                lastLinkedGlyph = (GroupedEffectGlyph) NextGlyph;
                while (lastLinkedGlyph.IsLinkedWithNext)
                {
                    lastLinkedGlyph = (GroupedEffectGlyph) lastLinkedGlyph.NextGlyph;
                }
            }

            // use start position of next beat when possible
            var endBeatRenderer = lastLinkedGlyph.Renderer;
            var endBeat = lastLinkedGlyph.Beat;
            var position = EndPosition;

            // calculate end X-position
            var cxRenderer = cx - Renderer.X;
            var endX = CalculateEndX(endBeatRenderer, endBeat, cxRenderer, position);

            PaintGrouped(cx, cy, endX, canvas);
        }

        protected virtual float CalculateEndX(BarRendererBase endBeatRenderer, Beat endBeat, float cx, BeatXPosition endPosition)
        {
            if (endBeat == null)
            {
                return cx + endBeatRenderer.X + X + Width;
            }
            else
            {
                return cx + endBeatRenderer.X + endBeatRenderer.GetBeatX(endBeat, endPosition);
            }
        }

        protected virtual void PaintNonGrouped(float cx, float cy, ICanvas canvas)
        {
            var cxRenderer = cx - Renderer.X;
            var endX = CalculateEndX(Renderer, Beat, cxRenderer, EndPosition);
            PaintGrouped(cx, cy, endX, canvas);
        }

        protected abstract void PaintGrouped(float cx, float cy, float endX, ICanvas canvas);
    }
}

File: src/rendering/glyphs/GroupedEffectGlyph.ts
import type { Beat } from '@src/model/Beat';
import type { ICanvas } from '@src/platform/ICanvas';
import type { BarRendererBase } from '@src/rendering/BarRendererBase';
import type { BeatXPosition } from '@src/rendering/BeatXPosition';
import { EffectGlyph } from '@src/rendering/glyphs/EffectGlyph';

export abstract class GroupedEffectGlyph extends EffectGlyph {
    protected endPosition: BeatXPosition;
    protected forceGroupedRendering: boolean = false;
    protected endOnBarLine: boolean = false;

    protected constructor(endPosition: BeatXPosition) {
        super();
        this.endPosition = endPosition;
    }

    public get isLinkedWithPrevious(): boolean {
        return !!this.previousGlyph && this.previousGlyph.renderer.staff.system === this.renderer.staff.system;
    }

    public get isLinkedWithNext(): boolean {
        return (
            !!this.nextGlyph &&
            this.nextGlyph.renderer.isFinalized &&
            this.nextGlyph.renderer.staff.system === this.renderer.staff.system
        );
    }

    public override paint(cx: number, cy: number, canvas: ICanvas): void {
        // if we are linked with the previous, the first glyph of the group will also render this one.
        if (this.isLinkedWithPrevious) {
            return;
        }
        // we are not linked with any glyph therefore no expansion is required, we render a simple glyph.
        if (!this.isLinkedWithNext && !this.forceGroupedRendering) {
            this.paintNonGrouped(cx, cy, canvas);
            return;
        }
        // find last linked glyph that can be
        let lastLinkedGlyph: GroupedEffectGlyph;
        if (!this.isLinkedWithNext && this.forceGroupedRendering) {
            lastLinkedGlyph = this;
        } else {
            lastLinkedGlyph = this.nextGlyph as GroupedEffectGlyph;
            while (lastLinkedGlyph.isLinkedWithNext) {
                lastLinkedGlyph = lastLinkedGlyph.nextGlyph as GroupedEffectGlyph;
            }
        }
        // use start position of next beat when possible
        const endBeatRenderer: BarRendererBase = lastLinkedGlyph.renderer;
        const endBeat: Beat = lastLinkedGlyph.beat!;
        const position: BeatXPosition = this.endPosition;
        // calculate end X-position
        const cxRenderer: number = cx - this.renderer.x;
        const endX: number = this.calculateEndX(endBeatRenderer, endBeat, cxRenderer, position);
        this.paintGrouped(cx, cy, endX, canvas);
    }

    protected calculateEndX(
        endBeatRenderer: BarRendererBase,
        endBeat: Beat | null,
        cx: number,
        endPosition: BeatXPosition
    ): number {
        if (!endBeat) {
            return cx + endBeatRenderer.x + this.x + this.width;
        }
        return cx + endBeatRenderer.x + endBeatRenderer.getBeatX(endBeat, endPosition);
    }

    protected paintNonGrouped(cx: number, cy: number, canvas: ICanvas): void {
        const cxRenderer: number = cx - this.renderer.x;
        const endX: number = this.calculateEndX(this.renderer, this.beat, cxRenderer, this.endPosition);
        this.paintGrouped(cx, cy, endX, canvas);
    }

    protected abstract paintGrouped(cx: number, cy: number, endX: number, canvas: ICanvas): void;
}
